File: src/main/java/com/emilianomenendez/veritran/bankaccount/Balance.java
package com.emilianomenendez.veritran.bankaccount;

import com.emilianomenendez.veritran.money.Money;
import lombok.Getter;
import lombok.ToString;

import java.util.Objects;

@Getter
@ToString
public abstract class Balance {
    private final String currency;
    private final int amount;

    public static Balance zero(String currency) {
        return create(currency, 0);
    }

    public static Balance create(Money initialBalance) {
        return Balance.create(initialBalance.getCurrency(), initialBalance.getAmount());
    }

    private static Balance create(String currency, int amount) {
        if (amount >= 0) {
            return new PositiveBalance(currency, amount);
        } else {
            return new NegativeBalance(currency, amount);
        }
    }

    public static Balance positive(Money amount) {
        return new PositiveBalance(amount.getCurrency(), amount.getAmount());
    }

    public static Balance negative(Money amount) {
        return new NegativeBalance(amount.getCurrency(), amount.getAmount());
    }

    public Balance(String currency, int amount) {
        this.currency = currency;
        this.amount = amount;
    }

    public boolean isGreaterThanOrEqual(Money other) {
        return amount >= other.getAmount();
    }

    public boolean isGreaterThanOrEqual(Balance other) {
        return getAmount() >= other.getAmount();
    }

    public Balance plus(Balance other) {
        return create(this.currency, this.amount + other.getAmount());
    }

    public Balance plus(Money other) {
        return create(this.currency, this.amount + other.getAmount());
    }

    public Balance minus(Money other) {
        return create(this.currency, this.amount - other.getAmount());
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (!(o instanceof Balance)) return false;
        Balance balance = (Balance) o;
        return amount == balance.amount && (amount == 0 || currency.equals(balance.currency));
    }

    @Override
    public int hashCode() {
        return Objects.hash(currency, amount);
    }
}

File: src/main/kotlin/ar/com/flow/bankaccount/domain/balance/Balance.kt
package ar.com.flow.bankaccount.domain.balance

import ar.com.flow.bankaccount.domain.Currency
import ar.com.flow.money.Money

class Balance(val currency: Currency, val amount: Int) {
    fun isGreaterThanOrEqual(other: Money): Boolean {
        return amount >= other.amount
    }

    fun isGreaterThanOrEqual(other: Balance): Boolean {
        return amount >= other.amount
    }

    operator fun plus(other: Balance): Balance {
        return create(currency, amount + other.amount)
    }

    operator fun plus(other: Money): Balance {
        return create(currency, amount + other.amount)
    }

    operator fun minus(other: Money): Balance {
        return create(currency, amount - other.amount)
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as Balance

        if (currency != other.currency) return false
        if (amount != other.amount) return false

        return true
    }

    override fun hashCode(): Int {
        var result = currency.hashCode()
        result = 31 * result + amount
        return result
    }

    override fun toString(): String {
        return "Balance(currency='$currency', amount=$amount)"
    }

    companion object {
        @JvmStatic
        fun zero(currency: Currency): Balance {
            return create(currency, 0)
        }

        fun create(initialBalance: Money): Balance {
            return create(initialBalance.currency, initialBalance.amount)
        }

        fun create(currency: Currency, amount: Int): Balance {
            return if (amount >= 0) {
                PositiveBalance.of(currency, amount)
            } else {
                NegativeBalance.of(currency, amount)
            }
        }

        @JvmStatic
        fun positive(amount: Money): Balance {
            return PositiveBalance.of(amount.currency, amount.amount)
        }

        @JvmStatic
        fun negative(amount: Money): Balance {
            return NegativeBalance.of(amount.currency, amount.amount)
        }
    }
}